src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+        private const int BatchMaxSaveIntervalSeconds = 37;
+
+        /// <summary>Maximum number of bytes the batch can hold until the downloaded blocks are stored to the disk.</summary>
+        private const uint BatchThresholdSizeBytes = 5 * 1000 * 1000;
+
+        /// <summary>The current batch size in bytes.</summary>
+        private int currentBatchSizeBytes;
+
+        /// <summary>The highest stored block in the repository.</summary>
+        public ChainedHeader StoreTip { get; private set; }
+
+        /// <summary>Items count in the <see cref="blocksQueue"/>.</summary>
+        internal int BlocksQueueCount => this.blocksQueue.Count;
+
+        /// <inheritdoc cref="ILogger"/>
+        private readonly ILogger logger;
+
+        /// <inheritdoc cref="INodeLifetime"/>
+        private readonly INodeLifetime nodeLifetime;
+
+        /// <inheritdoc cref="IChainState"/>
+        private readonly IChainState chainState;
+
+        /// <inheritdoc cref="StoreSettings"/>
+        private readonly StoreSettings storeSettings;
+
+        /// <inheritdoc cref="ConcurrentChain"/>
+        private readonly ConcurrentChain chain;
+
+        /// <inheritdoc cref="IBlockRepository"/>
+        private readonly IBlockRepository blockRepository;
+
+        /// <summary>Queue which contains blocks that should be saved to the database.</summary>
+        private readonly AsyncQueue<BlockPair> blocksQueue;
+
+        /// <summary>Task that runs <see cref="DequeueBlocksContinuouslyAsync"/>.</summary>
+        private Task dequeueLoopTask;
+
+        public BlockStore(
+            IBlockRepository blockRepository,
+            ConcurrentChain chain,
+            IChainState chainState,
+            StoreSettings storeSettings,
+            INodeLifetime nodeLifetime,
+            ILoggerFactory loggerFactory)
+        {
+            this.chainState = chainState;
+            this.nodeLifetime = nodeLifetime;
+            this.storeSettings = storeSettings;
+            this.chain = chain;
+            this.blockRepository = blockRepository;
+
+            this.blocksQueue = new AsyncQueue<BlockPair>();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="BlockStore"/>.
+        /// <para>
+        /// If StoreTip is <c>null</c>, the store is out of sync. This can happen when:</para>
+        /// <list>
+        ///     <item>The node crashed.</item>
+        ///     <item>The node was not closed down properly.</item>
+        /// </list>
+        /// <para>
+        /// To recover we walk back the chain until a common block header is found and set the <see cref="BlockStore"/>'s <see cref="StoreTip"/> to that.
+        /// </para>
+        /// </summary>
+        public async Task InitializeAsync()
+        {
+            this.logger.LogTrace("()");
+
+            if (this.storeSettings.ReIndex)
+                throw new NotImplementedException();
```

Comment:
? not implemented

Reply:
Yep. Also from prev implementation. 
It was never implemented on our full node.

Reply:
Ok, can the code actually get into that state? If so, perhaps we want NotSupportedException. If it can’t then the code redundant but acting like a todo I suppose.

Reply:
>can the code actually get into that state?

Yep

>If so, perhaps we want NotSupportedException. 

Well, it should be supported. Just not implemented.

Reply:
Should be supporrted, yes, but it isn't therefore the Exception could be correct as that.

But we don't want an end user to get Not Implemented Exception - that feels like it wasn't intentional.

Reply:
Fixed